src/main/scala/com/fulcrumgenomics/util/Metric.scala

```diff
@@ -102,53 +99,14 @@ object Metric extends LazyLogging {
   def iterator[T <: Metric](lines: Iterator[String], source: Option[String] = None)(implicit tt: ru.TypeTag[T]): Iterator[T] = {
-    val clazz: Class[T]   = ReflectionUtil.typeTagToClass[T]
-
-    def fail(lineNumber: Int,
-             message: String,
-             throwable: Option[Throwable] = None): Unit = {
-      val sourceMessage = source.map("\nIn source: " + _).getOrElse("")
-      val fullMessage   = s"On line #$lineNumber for metric '${clazz.getSimpleName}'$sourceMessage\n$message"
-      throwable.foreach { thr =>
-        val stringWriter = new StringWriter
-        thr.printStackTrace(new PrintWriter(stringWriter))
-        val banner = "#" * 80
-        logger.debug(banner)
-        logger.debug(stringWriter.toString)
-        logger.debug(banner)
-      }
-      throw FailureException(message=Some(fullMessage))
-    }
-
-    if (lines.isEmpty) fail(lineNumber=1, message="No header found")
-    val parser = new DelimitedDataParser(lines=lines, delimiter=Delimiter, ignoreBlankLines=false, trimFields=true)
-    val names  = parser.headers.toIndexedSeq
-    val reflectiveBuilder = new ReflectiveBuilder(clazz)
+    val builder = new MetricBuilder[T](source=source)(tt)
+    if (lines.isEmpty) builder.fail(message="No header found", lineNumber=Some(1))
+    val parser  = new DelimitedDataParser(lines=lines, delimiter=Delimiter, ignoreBlankLines=false, trimFields=true)
+    val names   = parser.headers.toIndexedSeq
 
     parser.zipWithIndex.map { case (row, rowIndex) =>
-      forloop(from = 0, until = names.length) { i =>
-        reflectiveBuilder.argumentLookup.forField(names(i)) match {
-          case Some(arg) =>
-            val value = {
-              val tmp = row[String](i)
-              if (tmp.isEmpty && arg.argumentType == classOf[Option[_]]) ReflectionUtil.SpecialEmptyOrNoneToken else tmp
-            }
-
-            val argumentValue = ReflectionUtil.constructFromString(arg.argumentType, arg.unitType, value) match {
-              case Success(v) => v
-              case Failure(thr) =>
-                fail(lineNumber=rowIndex+2, message=s"Could not construct value for column '${arg.name}' of type '${arg.typeDescription}' from '$value'", Some(thr))
-            }
-            arg.value = argumentValue
-          case None =>
-            fail(lineNumber=rowIndex+2, message=s"Did not have a field with name '${names(i)}'.")
-        }
-      }
-
-      // build it.  NB: if arguments are missing values, then an exception will be thrown here
-      // Also, we don't use the default "build()" method since if a collection or option is empty, it will be treated as
-      // missing.
-      reflectiveBuilder.build(reflectiveBuilder.argumentLookup.ordered.map(arg => arg.value getOrElse unreachable(s"Arguments not set: ${arg.name}")))
+      val argMap = names.zipWithIndex.map { case (name, i) => name -> row[String](i) }.toMap
+      builder.fromArgMap(argMap=argMap, lineNumber=Some(rowIndex+2))
```

Comment:
I'm now a tad confused why `fromArgMap` is a necessary method on the Metric builder. The MetricBuilder already knows what the field names are, so the keys to this map are not necessary and you can use a method as simple as `.fromValues()`.

Reply:
Since not all values need to specified, for example when the metric has default values.